File: project/dotnet_sol/MaMapF/Models/Constraint.cs
using System;

namespace MaMapF.Models
{
    public class Constraint
    {
        public char Agent { get; set; }
        public Position Position { get; set; }

        public int Step { get; set; }

        public override string ToString()
        {
            return $"CONSTRAINT\nAgent: {Agent}, Position: {Position}, Step: {Step}";
        }

        public bool Equals(Constraint other)
        {
            return Agent == other.Agent && Position.Equals(other.Position) && Step == other.Step;
        }

        public override bool Equals(object? obj)
        {
            if (obj == null) return false;
            if (!(obj is Constraint other)) return false;
            return Equals(other);
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(Agent, Position, Step);
        }

        public Constraint Copy()
        {
            return new Constraint
            {
                Agent = Agent,
                Position = Position.Next(0, 0),
                Step = Step + 0,
            };
        }
    }
}

File: project/python_sol/src/models/constraint.py
from src.models.conflict import Conflict


class Constraint:
    def __init__(self, agent: str, position: [int, ...], step: int, conflict: Conflict):
        self.agent: str = agent
        self.position: [int, ...] = position
        self.step: int = step
        self.conflict: Conflict = conflict

    def __repr__(self):
        return "CONSTRAINT: agent: {}, position: {}, step: {}\n" \
            .format(self.agent, self.position, self.step)

    def __eq__(self, other):
        return self.step == other.step \
               and self.agent == other.agent \
               and self.position == other.position
